File: sims-market/server/order_book.go
package main

import "sync"

type OrderBook struct {
  Bids *List[Order] `json:"bids"`
  Asks *List[Order] `json:"asks"`
  /*
  MarketBuys []Order `json:"marketBuys"`
  MarketSells []Order `json:"marketSells"`
  */
  MarketBuys *List[Order] `json:"marketBuys"`
  MarketSells *List[Order] `json:"marketSells"`
  mtx sync.RWMutex
}

func NewOrderBook() *OrderBook {
  return &OrderBook{
    Bids: NewList[Order](),
    Asks: NewList[Order](),
    MarketBuys: NewList[Order](),
    MarketSells: NewList[Order](),
  }
}

func (ob *OrderBook) addBid(order Order) {
  ob.mtx.Lock()
  ob.lockedAddBid(order)
  ob.mtx.Unlock()
}

func (ob *OrderBook) lockedAddBid(order Order) {
  // Check market orders
  order = ob.lockedCheckMarketSells(order)
  // TODO: Notify?
  if order.IsFilled() {
    return
  }
  // Loop through finding asks with prices less than or equal to the order
  // price
  for n := ob.Asks.Head(); n != nil && n.Value.Price <= order.Price; n = n.Next() {
    // TODO: Update filled price
    // TODO: Notify on order book change
    ask := &n.Value
    order.FillWith(ask)
    if ask.IsFilled() {
      n.Remove()
    }
    if order.IsFilled() {
      return
    }
  }
  for n := ob.Bids.Head(); n != nil; n = n.Next() {
    if n.Value.Price < order.Price {
      n.InsertBefore(order)
      return
    }
  }
  ob.Bids.PushBack(order)
}

func (ob *OrderBook) addMarketBuy(order Order) {
  ob.mtx.Lock()
  ob.lockedAddMarketBuy(order)
  ob.mtx.Unlock()
}

func (ob *OrderBook) lockedAddMarketBuy(order Order) {
  if !ob.MarketBuys.IsEmpty() {
    ob.MarketBuys.PushBack(order)
    return
  }
  ob.Asks.Range(func(n *Node[Order]) bool {
    // TODO: Notify
    ask := &n.Value
    filledSize := ask.Fill(ask.Price, order.Size)
    if ask.IsFilled() {
      n.Remove()
    }
    order.Fill(ask.Price, filledSize)
    return !order.IsFilled()
  })
  if !order.IsFilled() {
    // NOTE: MarketBuys will be empty
    ob.MarketBuys.PushFront(order)
  }
}

func (ob *OrderBook) lockedCheckMarketBuys(order Order) Order {
  // TODO: Notify
  ob.MarketBuys.Range(func(n *Node[Order]) bool {
    order.FillWith(&n.Value)
    return !order.IsFilled()
  })
  return order
}

func (ob *OrderBook) addAsk(order Order) {
  ob.mtx.Lock()
  ob.lockedAddAsk(order)
  ob.mtx.Unlock()
}

func (ob *OrderBook) lockedAddAsk(order Order) {
  // Check market orders
  order = ob.lockedCheckMarketBuys(order)
  // TODO: Notify?
  if order.IsFilled() {
    return
  }
  // Loop through finding bids with prices greater than or equal to the order
  // price
  for n := ob.Bids.Head(); n != nil && n.Value.Price >= order.Price; n = n.Next() {
    // TODO: Update filled price
    // TODO: Notify on order book change
    bid := &n.Value
    order.FillWith(bid)
    if bid.IsFilled() {
      n.Remove()
    }
    if order.IsFilled() {
      return
    }
  }
  for n := ob.Asks.Head(); n != nil; n = n.Next() {
    if n.Value.Price > order.Price {
      n.InsertBefore(order)
      return
    }
  }
  ob.Asks.PushBack(order)
}

func (ob *OrderBook) addMarketSell(order Order) {
  ob.mtx.Lock()
  ob.lockedAddMarketSell(order)
  ob.mtx.Unlock()
}

func (ob *OrderBook) lockedAddMarketSell(order Order) {
  if !ob.MarketSells.IsEmpty() {
    ob.MarketSells.PushBack(order)
    return
  }
  ob.Bids.Range(func(n *Node[Order]) bool {
    // TODO: Notify
    bid := &n.Value
    filledSize := bid.Fill(bid.Price, order.Size)
    if bid.IsFilled() {
      n.Remove()
    }
    order.Fill(bid.Price, filledSize)
    return order.IsFilled()
  })
  if !order.IsFilled() {
    // NOTE: MarketSells will be empty
    ob.MarketSells.PushFront(order)
  }
}

func (ob *OrderBook) lockedCheckMarketSells(order Order) Order {
  // TODO: Notify
  ob.MarketSells.Range(func(n *Node[Order]) bool {
    order.FillWith(&n.Value)
    return !order.IsFilled()
  })
  return order
}

type OrderType int

const (
  OrderTypeUnknown OrderType = iota
  OrderTypeMarket
  OrderTypeLimit
)

type OrderSide int

const (
  OrderSideUnknown OrderSide = iota
  OrderSideBuy
  OrderSideSell
)

// TODO: Size things
type Order struct {
  Id uint64 `json:"id"`
  User string `json:"user"`

  Type OrderType `json:"orderType"`
  Side OrderSide `json:"side"`

  Price float64 `json:"price"`
  Size int64 `json:"size"`

  FilledPrice float64 `json:"filledPrice"`
  FilledSize int64 `json:"filledSize"`

  Canceled bool `json:"canceled"`
}

func (o Order) RemainingSize() int64 {
  return o.Size - o.FilledSize
}

// Fill attempts to fill the order with the given price and size. This method
// does not check to make sure order types and prices match, that is the job of
// the caller. It returns the size filled, which will be <= size (less if the
// remaining size is less than the passed size).
func (o *Order) Fill(price float64, size int64) int64 {
  if r := o.RemainingSize(); r < size {
    size = r
  }
  ffilled, fsize := float64(o.FilledSize), float64(size)
  fNewFilled := ffilled + fsize
  oldPart := o.FilledPrice / fNewFilled * ffilled
  newPart := price / fNewFilled * fsize
  o.FilledPrice = oldPart + newPart
  o.FilledSize += size
  return size
}

// FillWith fills the calling order with the passed order as much as possible,
// using the passed order's price and size. It will also fill the passed order,
// calling Fill on both. It returns the size filled, which will be
// <= order.Size. See Order.Fill for further detail.
func (o *Order) FillWith(order *Order) int64 {
  filled := o.Fill(order.Price, order.Size)
  order.Fill(order.Price, filled)
  return filled
}

func (o Order) IsFilled() bool {
  return o.RemainingSize() == 0
}

File: sims-market/desktop/src/order_book.rs
#![allow(dead_code)]
use std::collections::LinkedList;

pub struct OrderBook {
    bids: LinkedList<Order>,
    asks: LinkedList<Order>,
    market_buys: LinkedList<Order>,
    market_sells: LinkedList<Order>,
}

impl OrderBook {
    pub fn new() -> Self {
        Self {
            bids: LinkedList::new(),
            asks: LinkedList::new(),
            market_buys: LinkedList::new(),
            market_sells: LinkedList::new(),
        }
    }

    pub fn bids(&self) -> &LinkedList<Order> {
        &self.bids
    }

    pub fn asks(&self) -> &LinkedList<Order> {
        &self.asks
    }

    pub fn market_buys(&self) -> &LinkedList<Order> {
        &self.market_buys
    }

    pub fn market_sells(&self) -> &LinkedList<Order> {
        &self.market_sells
    }

    pub fn add_bid(&mut self, mut order: Order) {
        self.check_market_sells(&mut order);
        if order.is_filled() {
            return;
        }

        let mut cur = self.asks.cursor_front_mut();
        while let Some(ask) = cur.current() {
            if ask.price > order.price {
                break;
            }
            order.fill_with(ask);
            if ask.is_filled() {
                cur.remove_current();
            } else {
                cur.move_next();
            }
            if order.is_filled() {
                return;
            }
        }
        let mut cur = self.bids.cursor_front_mut();
        while let Some(bid) = cur.current() {
            if bid.price < order.price {
                cur.insert_before(order);
                return;
            }
            cur.move_next();
        }
        self.bids.push_back(order);
    }

    pub fn add_market_buy(&mut self, mut order: Order) {
        if !self.market_buys.is_empty() {
            self.market_buys.push_back(order);
            return;
        }
        let mut cur = self.asks.cursor_front_mut();
        while let Some(ask) = cur.current() {
            let filled = ask.fill(ask.price, order.size);
            order.fill(ask.price, filled);
            if order.is_filled() {
                return;
            }
            if ask.is_filled() {
                cur.remove_current();
            } else {
                cur.move_next();
            }
        }
        self.market_buys.push_back(order);
    }

    fn check_market_buys(&mut self, order: &mut Order) {
        let mut cur = self.market_buys.cursor_front_mut();
        while let Some(buy) = cur.current() {
            order.fill_with(buy);
            if order.is_filled() {
                return;
            }
            cur.move_next();
        }
    }

    pub fn add_ask(&mut self, mut order: Order) {
        self.check_market_sells(&mut order);
        if order.is_filled() {
            return;
        }

        let mut cur = self.bids.cursor_front_mut();
        while let Some(bid) = cur.current() {
            if bid.price < order.price {
                break;
            }
            order.fill_with(bid);
            if order.is_filled() {
                return;
            }
            if bid.is_filled() {
                cur.remove_current();
            } else {
                cur.move_next();
            }
        }
        let mut cur = self.asks.cursor_front_mut();
        while let Some(ask) = cur.current() {
            if ask.price > order.price {
                cur.insert_before(order);
                return;
            }
            cur.move_next();
        }
        self.asks.push_back(order);
    }

    pub fn add_market_sell(&mut self, mut order: Order) {
        if !self.market_sells.is_empty() {
            self.market_sells.push_back(order);
            return;
        }
        let mut cur = self.bids.cursor_front_mut();
        while let Some(bid) = cur.current() {
            let filled = bid.fill(bid.price, order.size);
            order.fill(bid.price, filled);
            if order.is_filled() {
                return;
            }
            if bid.is_filled() {
                cur.remove_current();
            } else {
                cur.move_next();
            }
        }
        self.market_sells.push_back(order);
    }

    fn check_market_sells(&mut self, order: &mut Order) {
        let mut cur = self.market_sells.cursor_front_mut();
        while let Some(sell) = cur.current() {
            order.fill_with(sell);
            if order.is_filled() {
                return;
            }
            cur.move_next();
        }
    }
}

#[derive(Default, Clone)]
pub struct Order {
    id: u64,
    user_id: u64,

    pub otype: OrderType,
    pub side: OrderSide,

    pub price: f64,
    pub size: u64,

    filled_price: f64,
    filled_size: u64,

    canceled: bool,
}

impl Order {
    pub fn otype(mut self, otype: OrderType) -> Self {
        self.otype = otype;
        self
    }

    pub fn side(mut self, side: OrderSide) -> Self {
        self.side = side;
        self
    }

    pub fn price(mut self, price: f64) -> Self {
        self.price = price;
        self
    }

    pub fn size(mut self, size: u64) -> Self {
        self.size = size;
        self
    }

    pub fn filled_price(&self) -> f64 {
        self.filled_price
    }

    pub fn filled_size(&self) -> u64 {
        self.filled_size
    }

    pub fn canceled(&self) -> bool {
        self.canceled
    }

    pub fn remaining_size(&self) -> u64 {
        self.size - self.filled_size
    }

    fn fill(&mut self, price: f64, size: u64) -> u64 {
        let r = self.remaining_size();
        let size = if r < size { r } else { size };
        let (ffilled, fsize) = (self.filled_size as f64, size as f64);
        let f_new_filled = ffilled + fsize;
        let old_part = self.filled_price / f_new_filled * ffilled;
        let new_part = price / f_new_filled * fsize;
        self.filled_price = old_part + new_part;
        self.filled_size += size;
        size
    }

    fn fill_with(&mut self, order: &mut Order) -> u64 {
        let filled = self.fill(order.price, order.size);
        order.fill(order.price, filled);
        filled
    }

    pub fn is_filled(&self) -> bool {
        self.size == self.filled_size
    }
}

#[derive(Clone, Copy, Default, PartialEq, Eq)]
pub enum OrderType {
    #[default]
    Unknown,
    Limit,
    Market,
}

#[derive(Clone, Copy, Default, PartialEq, Eq)]
pub enum OrderSide {
    #[default]
    Unknown,
    Buy,
    Sell,
}
